vmdb/app/models/mixins/ems_openstack_mixin.rb

```diff
@@ -0,0 +1,150 @@
+      password = options[:pass] || self.authentication_password(options[:auth_type])
+      osh = OpenstackHandle::Handle.new(username, password, address, port)
+      osh.connection_options = {:instrumentor => $fog_log}
+      osh
+    end
+  end
+
+  def reset_openstack_handle
+    @openstack_handle = nil
+  end
+
+  def connect(options = {})
+    openstack_handle(options).connect(options)
+  end
+
+  def connect_volume
+    connect(:service => "Volume")
+  end
+
+  def connect_identity
+    connect(:service => "Identity")
+  end
+
+  def event_monitor_options
+    @event_monitor_options ||= begin
+      opts = {:hostname => self.ipaddress}
+      opts[:port] = MiqEventCatcherOpenstack.worker_settings[:amqp_port]
+      if self.has_authentication_type? :amqp
+        # authentication_userid/password will happily return the "default"
+        # userid/password if this ems has no amqp auth configured
+        opts[:username] = self.authentication_userid(:amqp)
+        opts[:password] = self.authentication_password(:amqp)
+      end
+      opts
+    end
+  end
+
+  def event_monitor_available?
+    require 'openstack/openstack_event_monitor'
+    OpenstackEventMonitor.available?(event_monitor_options)
+  rescue => e
+    $log.error("MIQ(#{self.class.name}.#{__method__}) Exeption trying to find openstack event monitor for #{self.name}(#{self.hostname}). #{e.message}")
+    false
+  end
+
+  def verify_api_credentials(options={})
+    begin
+      with_provider_connection(options) {}
+    rescue Excon::Errors::Unauthorized => err
+      $log.error("MIQ(#{self.class.name}.verify_api_credentials) Error Class=#{err.class.name}, Message=#{err.message}")
+      raise MiqException::MiqEVMLoginError, "Login failed due to a bad username or password."
+    rescue Exception => err
+      $log.error("MIQ(#{self.class.name}.verify_api_credentials) Error Class=#{err.class.name}, Message=#{err.message}")
+      raise MiqException::MiqEVMLoginError, "Unexpected response returned from system, see log for details"
+    end
+    true
+  end
+  private :verify_api_credentials
+
+  def verify_amqp_credentials(options={})
+    require 'openstack/openstack_event_monitor'
+    OpenstackEventMonitor.test_amqp_connection(event_monitor_options)
+  rescue Exception => e
+    $log.error("MIQ(#{self.class.name}.verify_amqp_credentials) Error Class=#{e.class.name}, Message=#{e.message}")
+    raise MiqException::MiqEVMLoginError, e.to_s
+  end
+  private :verify_amqp_credentials
+
+  def verify_credentials(auth_type=nil, options={})
+    auth_type ||= 'default'
+
+    raise MiqException::MiqHostError, "No credentials defined" if self.authentication_invalid?(auth_type)
+
+    options.merge!(:auth_type => auth_type)
+    case auth_type.to_s
+    when 'default'; verify_api_credentials(options)
+    when 'amqp';    verify_amqp_credentials(options)
+    else;           raise "Invalid OpenStack Authentication Type: #{auth_type.inspect}"
+    end
+  end
+
+  #
+  # Operations
+  #
```

Comment:
I think the VM Operations below should be left in the `EmsOpenstack` (cloudy) class.